### DISCUSSION TITLE
Can I Use Separate Alembic Migrations for Two Services Sharing the Same Database?

### DISCUSSION THREAD

**Discussion by Saqibs575** [sqlalchemy/alembic#1522] [Usage Questions]
I have two different services that are using the same database. I want to manage schema migrations separately for each service, using Alembic in their respective virtual environments. Is it possible to initialize and maintain separate Alembic migrations for both services while sharing the same database?

If it’s not feasible, what would be the recommended approach for managing migrations in this scenario?

Note that both alembic migrations should be independent.

**Comment by CaselIT** [comment]
Hi,

Are the two services using the same schema? if not configure alembic to look only at the relevant schema
If they use the same schema, are they sharing the same tables? if not configure alembic to look only for the relevant tables

If the two services share both the schema and the table I would suggest you to rethink your schema. Or use a 3rd project to manage the common schema of both services.

**Reply by Saqibs575** [reply]
I have initialized migrations separately using Alembic in the respective virtual environments of each service. Each service is intended to handle its own migrations independently. However, I’m encountering the following issue when creating a new revision for the second service:

```vbnet
alembic revision --autogenerate -m "initial migration2"
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
ERROR [alembic.util.messaging] Can't locate revision identified by 'ae12b966b0f4'
FAILED: Can't locate revision identified by 'ae12b966b0f4'
```
The revision id ` 'ae12b966b0f4'` belongs to revision of first service migration.
I have configured the env.py file to include only specific tables in the migration context with the following code:

```python
def include_object(object, name, type_, reflected, compare_to):
    # Example: Only include tables 'test_model_1' and 'test_model_2'
    if type_ == 'table' and name in ['test_model_1', 'test_model_2']:
        return True
    return False
    
def run_migrations_offline() -> None:
    """Run migrations in 'offline' mode.

    This configures the context with just a URL
    and not an Engine, though an Engine is acceptable
    here as well.  By skipping the Engine creation
    we don't even need a DBAPI to be available.

    Calls to context.execute() here emit the given string to the
    script output.

    """
    url = config.get_main_option("sqlalchemy.url")
    context.configure(
        url=url,
        target_metadata=target_metadata,
        literal_binds=True,
        dialect_opts={"paramstyle": "named"}, include_object=include_object
    )

    with context.begin_transaction():
        context.run_migrations()


def run_migrations_online() -> None:
    """Run migrations in 'online' mode.

    In this scenario we need to create an Engine
    and associate a connection with the context.

    """
    connectable = engine_from_config(
        config.get_section(config.config_ini_section, {}),
        prefix="sqlalchemy.",
        poolclass=pool.NullPool,
    )

    with connectable.connect() as connection:
        context.configure(
            connection=connection, target_metadata=target_metadata, include_object=include_object)

        with context.begin_transaction():
            context.run_migrations()

```
Despite this configuration, the error persists. How can I address this issue, considering that both services should maintain independent migration histories while sharing the same database?

**Reply by CaselIT** [reply]
Forgot to mention that the two alembic env must have a different alembic history table. Specify if using https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.configure.params.version_table

**Reply by Saqibs575** [reply]
@CaselIT Thank you very much.
